### PR TITLE
[Settings] Handle `woocommerce_settings_xxx` action for custom views

### LINKED ISSUE #52317: [Settings] Custom views - Handle `woocommerce_settings_xxx` action
While looping through Settings pages to construct the JSON that powers Settings, handle custom views created by the `woocommerce_settings_xxx` action.

https://github.com/woocommerce/woocommerce/blob/2d65567a6ad261a043a95b4f758f72ed190fd0de/plugins/woocommerce/includes/admin/views/html-admin-settings.php#L53

For reference, this is how the Proof of concept handled this situation.

https://github.com/woocommerce/woocommerce/blob/09cdb35a1ebbc422f87c5dea1795b0b7824d254e/plugins/woocommerce/includes/admin/settings/class-wc-settings-page.php#L147-L154

Note that extensions often overwrite the `output` function with custom views that also call `parent::output` as well. This case must also be handled.

Add tests as well

### PR DESCRIPTION
**Pull request by chihsuan**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes https://github.com/woocommerce/woocommerce/issues/52317.

This PR ensures that custom views created by the `woocommerce_settings_xxx action` are correctly handled while looping through WooCommerce Settings pages to construct the JSON data that powers the new Settings screens. The implementation accounts for extensions that overwrite the output function while still calling `parent::output`.

Changes:

1. **Custom View Handling**:
   - Added support for handling custom views using the `woocommerce_settings_xxx` action during page data generation for Settings.

2. **Parent Output Handling**:
   - Do not output anything when the feature flag is enabled.
   - Do not loop through the settings object when the parent output is NOT called.

3. **Unit Tests**:
   - Added tests to validate:
     - Proper handling of `woocommerce_settings_xxx` actions.
     - Parent output is called / not called.
     - JSON output consistency when custom views are injected.

---

I noticed that when tax is disabled, we still show the tax tab ionthe settings page. @psealock  Should we use the `apply_filters( 'woocommerce_settings_tabs_array', array() );` to check what tabs are available? However, if user enables tax later, we will need to refresh the setting data to show the tax tab. 🤔 

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Create a new JN site with Gutenberg, WC Beta Tester enabled
2. Go to `WooCommerce > Settings > General`
3. Check the `Enable tax rates and calculations` checkbox to output tax settings

https://github.com/woocommerce/woocommerce/blob/fcd67239e19775dcc124abbec4a2160f6ef0d557/plugins/woocommerce/includes/admin/settings/class-wc-settings-tax.php#L105-L119

4. Go to `/wp-admin/tools.php?page=woocommerce-admin-test-helper` > Features and enable `settings` feature flag
5. Go to `WooCommerce > Settings`
6. In the console, look for the tax's settings:

```js
window.wcSettings.admin.settingsData.tax.sections.default.settings

window.wcSettings.admin.settingsData.tax.sections.standard.settings[0].content
```

7. Observe that the default section contains the Tax options setting object

Screenshot 2024-12-18 at 09 44 40

8. Observe that the standard / other tax classes sections contain custom HTML content below:
Screenshot 2024-12-18 at 09 34 12

https://github.com/woocommerce/woocommerce/blob/fcd67239e19775dcc124abbec4a2160f6ef0d557/plugins/woocommerce/includes/admin/settings/views/html-settings-tax.php#L13-L155

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>

### CODE REVIEW

**Review by rjchow — APPROVED** [review]
Tested well, bit hard to understand what's going on

Not sure if its in the scope of this PR but we are going to have some trouble with WooCommerce Shipping:

image

This looks like it's coming from: https://github.com/woocommerce/woocommerce-shipping/blob/trunk/classes/class-wc-connect-settings-pages.php#L163

Edit:

Actually maybe we won't, it looks like the React script will pick it up properly if we just output the div with the id:
https://github.com/woocommerce/woocommerce-shipping/blob/trunk/client/entrypoints/shipping-settings/index.js

**Comment by chihsuan** [comment]
> Not sure if its in the scope of this PR but we are going to have some trouble with WooCommerce Shipping:
> Actually maybe we won't, it looks like the React script will pick it up properly if we just output the div with the id: https://github.com/woocommerce/woocommerce-shipping/blob/trunk/client/entrypoints/shipping-settings/index.js

Thanks for the info! I confirmed that the classic wc shipping setting page also renders the same HTML, and yes, the React script will pick it up properly.

Screenshot 2024-12-18 at 15 37 49

**Review by psealock — APPROVED** [review]
This is looking great @chihsuan. I also tested with the Settings Tester whose output method looks something like this:

```php
public function output() {
	global $current_section;

	// Render configured output method.
	parent::output();

	if ( 'custom_views' === $current_section ) {
		?>
		<h3 style="background-color: yellow; padding: 10px;">
			This html has been generated by PHP using a custom the `output` method.
		</h3>
		<?php
	}
}
```
I'm correctly seeing the output

image

It is a bit hard to follow but I think once we have https://github.com/woocommerce/woocommerce/issues/52973 displaying custom output it will be easier to spot any gotchas. So I think we should merge this and continue testing once the picture is more complete
